base/shell/explorer/traydeskbtn.cpp

```diff
@@ -0,0 +1,382 @@
+    Create(hwndParent, NULL, NULL, style);
+
+    if (!m_hWnd)
+        return E_FAIL;
+
+    // Get desktop icon
+    bool bIconRetrievalFailed = ExtractIconExW(L"imageres.dll", -IDI_IMAGERES_DESKTOP, NULL, &m_icon, 1) == UINT_MAX;
+    if (bIconRetrievalFailed || !m_icon)
+        ExtractIconExW(L"shell32.dll", -IDI_SHELL32_DESKTOP, NULL, &m_icon, 1);
+
+    // Get appropriate size at which to display desktop icon
+    m_szIcon.cx = GetSystemMetrics(SM_CXSMICON);
+    m_szIcon.cy = GetSystemMetrics(SM_CYSMICON);
+
+    // Prep visual style
+    EnsureWindowTheme(TRUE);
+
+    // Get HWND of Taskbar
+    m_hWndTaskbar = ::GetParent(hwndParent);
+    if (!::IsWindow(m_hWndTaskbar))
+        return E_FAIL;
+
+    return S_OK;
+}
+
+LRESULT CTrayShowDesktopButton::OnClick(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL& bHandled)
+{
+    // The actual action can be delayed as an expected behaviour.
+    // But a too late action is an unexpected behaviour.
+    LONG nTime0 = m_nClickedTime;
+    LONG nTime1 = ::GetMessageTime();
+    if (nTime1 - nTime0 >= 600) // Ignore after 0.6 sec
+        return 0;
+
+    // Show/Hide Desktop
+    ::SendMessage(m_hWndTaskbar, WM_COMMAND, TRAYCMD_TOGGLE_DESKTOP, 0);
+
+    return 0;
+}
+
+// This function is called from OnLButtonDown and parent.
+VOID CTrayShowDesktopButton::Click()
+{
+    // The actual action can be delayed as an expected behaviour.
+    m_nClickedTime = ::GetMessageTime();
+    PostMessage(TSDB_CLICK, 0, 0);
+}
+
+LRESULT CTrayShowDesktopButton::OnLButtonUp(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL& bHandled)
+{
+    m_bPressed = FALSE;
+    ReleaseCapture();
+    Invalidate(TRUE);
+
+    POINT pt;
+    ::GetCursorPos(&pt);
+    if (PtInButton(&pt))
+        Click(); // Left-click
+    return 0;
+}
+
+LRESULT CTrayShowDesktopButton::OnLButtonDown(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL& bHandled)
+{
+    m_bPressed = TRUE;
+    SetCapture();
+    Invalidate(TRUE);
+    return 0;
+}
+
+LRESULT CTrayShowDesktopButton::OnSettingChanged(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL& bHandled)
+{
+    LRESULT ret = OnThemeChanged(uMsg, wParam, lParam, bHandled);
+    EnsureWindowTheme(TRUE);
+    return ret;
+}
+
+LRESULT CTrayShowDesktopButton::OnThemeChanged(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL& bHandled)
+{
+    HIGHCONTRAST hcInfo;
+    hcInfo.cbSize = sizeof(hcInfo);
+    if (SystemParametersInfo(SPI_GETHIGHCONTRAST, sizeof(hcInfo), &hcInfo, FALSE))
+        m_highContrastMode = (hcInfo.dwFlags & HCF_HIGHCONTRASTON);
+
+    if (m_hTheme)
+    {
+        ::CloseThemeData(m_hTheme);
+        m_hTheme = NULL;
+    }
+    if (m_hFallbackTheme)
+    {
+        ::CloseThemeData(m_hFallbackTheme);
+        m_hFallbackTheme = NULL;
+    }
+
+    EnsureWindowTheme(FALSE);
+
+    Invalidate(TRUE);
+    return 0;
+}
+
+LRESULT CTrayShowDesktopButton::OnWindowPosChanged(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL& bHandled)
+{
+    EnsureWindowTheme(TRUE);
```

Comment:
Why does this need to be called in the window-position changed message handler?

Reply:
vertical taskbar

Reply:
oki